python/ray/autoscaler/local/node_provider.py

```diff
@@ -90,12 +111,75 @@ class LocalNodeProvider(NodeProvider):
+            provider_config, cluster_name = self._get_node_configs(
+                provider_config, cluster_name)
+
         NodeProvider.__init__(self, provider_config, cluster_name)
-        self.state = ClusterState("/tmp/cluster-{}.lock".format(cluster_name),
-                                  "/tmp/cluster-{}.state".format(cluster_name),
-                                  provider_config)
+        self.state = ClusterState(
+            "/tmp/cluster-{}.lock".format(cluster_name),
+            "/tmp/cluster-{}.state".format(cluster_name),
+            provider_config,
+        )
+
+    def _get_node_configs(self, provider_config, cluster_name):
+        # Concatenate user_id to make cluster names unique between users.
+        new_cluster_name = cluster_name + "_" + getpass.getuser()
+        request = {
+            "request_type": "get_node_ips",
+            "cluster_name": new_cluster_name,
+            "num_workers": provider_config["max_workers"],
+        }
+        node_ips = self.get_http_response(request, self.server_address)
+        if "head_ip" not in node_ips:
+            raise OSError("Not enough nodes are available." +
+                          " Consider requesting less workers.")
+        else:
+            provider_config["head_ip"] = node_ips["head_ip"]
+            provider_config["worker_ips"] = node_ips["worker_ips"]
+        return provider_config, new_cluster_name
+
+    @staticmethod
+    def get_http_response(request, server_address):
+        headers = {
+            "Content-Type": "application/json",
+        }
+        request_message = json.dumps(request).encode()
+        http_server_address = "http://" + server_address
+        try:
+            r = requests.get(
+                http_server_address,
+                data=request_message,
+                headers=headers,
+                timeout=None,
+            )
+        except (RemoteDisconnected, ConnectionError):
+            logger.error(
+                "Could not connect to: " + server_address +
+                ". Did you run ray onprem-server start <on-prem-server.yaml>?")
+            raise
+
+        response = r.json()
+        return response
+
+    def _still_valid_cluster(self):
+        if self.auto_scheduling_mode:
+            request = {
+                "request_type": "still_valid_cluster",
+                "cluster_name": self.cluster_name,
+                "provider_config": self.provider_config,
+            }
+            is_valid = self.get_http_response(request, self.server_address)
+            if not is_valid:
+                raise OSError(
+                    "Cluster: " + self.cluster_name +
+                    " is not valid any more. Please restart the cluster.")
 
     def non_terminated_nodes(self, tag_filters):
+        self._still_valid_cluster()
         workers = self.state.get()
         matching_ips = []
         for worker_ip, info in workers.items():
```

Comment:
I'm pretty confused about how this is working, shouldn't all node provider calls be going to the coordinator when it is enabled?

Reply:
Yes. But for fault tolerance, I didn't want the clusters to die when the server dies...

Reply:
but based in that comment, I assume it is fine.